File: clients/FSharp/Model/Action.fs
#nowarn "0058"
namespace Aicup2020.Model
type Action = {
    EntityActions: Map<int, EntityAction>;
    } with
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.EntityActions.Count
        this.EntityActions |> Map.iter (fun key value ->
            writer.Write key
            value.writeTo writer
        )
    static member readFrom(reader: System.IO.BinaryReader) = {
        EntityActions = [for _ in 1 .. reader.ReadInt32() do
            let key = reader.ReadInt32()
            let value = EntityAction.readFrom reader
            yield (key, value)
            ] |> Map.ofList
    }

File: clients/CSharp/Model/Action.cs
namespace Aicup2020.Model
{
    public struct Action
    {
        public System.Collections.Generic.IDictionary<int, Model.EntityAction> EntityActions { get; set; }
        public Action(System.Collections.Generic.IDictionary<int, Model.EntityAction> entityActions)
        {
            this.EntityActions = entityActions;
        }
        public static Action ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new Action();
            int EntityActionsSize = reader.ReadInt32();
            result.EntityActions = new System.Collections.Generic.Dictionary<int, Model.EntityAction>(EntityActionsSize);
            for (int i = 0; i < EntityActionsSize; i++)
            {
                int EntityActionsKey;
                EntityActionsKey = reader.ReadInt32();
                Model.EntityAction EntityActionsValue;
                EntityActionsValue = Model.EntityAction.ReadFrom(reader);
                result.EntityActions.Add(EntityActionsKey, EntityActionsValue);
            }
            return result;
        }
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(EntityActions.Count);
            foreach (var EntityActionsEntry in EntityActions)
            {
                var EntityActionsKey = EntityActionsEntry.Key;
                var EntityActionsValue = EntityActionsEntry.Value;
                writer.Write(EntityActionsKey);
                EntityActionsValue.WriteTo(writer);
            }
        }
    }
}
